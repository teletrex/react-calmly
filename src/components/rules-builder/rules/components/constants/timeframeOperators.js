/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import { RULE_INPUT_TYPES } from '../../../constants';

const { NONE, RANGE, SINGLE_VALUE, WHEN_RELATION, WITHIN_PERIOD } = RULE_INPUT_TYPES;

export const TIMEFRAME_OPERATORS = [
  // Within:
  {
    inputType: WITHIN_PERIOD,
    label: 'within the last',
    value: 'WITHIN_THE_LAST',
  },

  // Today:
  {
    inputType: NONE,
    label: 'on today',
    value: 'ON_TODAY',
  },
  {
    inputType: NONE,
    label: 'before today',
    value: 'BEFORE_TODAY',
  },

  // Any time:
  {
    inputType: NONE,
    label: 'at any time',
    value: 'AT_ANY_TIME',
  },

  // Date:
  {
    inputType: SINGLE_VALUE,
    label: 'on',
    value: 'ON',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'after',
    value: 'AFTER',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'before',
    value: 'BEFORE',
  },

  // Date range:
  {
    inputType: RANGE,
    label: 'between',
    value: 'BETWEEN',
  },

  // When relation:
  {
    inputType: WHEN_RELATION,
    label: 'exactly',
    value: 'EXACTLY',
  },
  {
    inputType: WHEN_RELATION,
    label: 'more than',
    value: 'MORE_THAN',
  },
  {
    inputType: WHEN_RELATION,
    label: 'less than',
    value: 'LESS_THAN',
  },
];
