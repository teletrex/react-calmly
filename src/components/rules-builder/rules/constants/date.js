/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




export const DATE_PATTERN = 'd{1,2}/d{1,2}/d{4}';
