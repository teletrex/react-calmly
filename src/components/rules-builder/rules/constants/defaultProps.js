/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




export const DEFAULT_PROPS = {
  INPUT_FIELD_VALIDATION: [],
  OPERATORS: [],
  TIMEFRAME: {},
};
