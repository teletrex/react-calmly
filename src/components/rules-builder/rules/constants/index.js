/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




export * from './date';
export * from './defaultProps';
export * from './periods';
export * from './propTypes';
export * from './ruleOperators';
export * from './validationMessages';
export * from './values';
export * from './whenRelations';
