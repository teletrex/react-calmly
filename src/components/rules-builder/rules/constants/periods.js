/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




export const PERIODS = [
  {
    label: 'days',
    value: 'DAYS',
  },
  {
    label: 'weeks',
    value: 'WEEKS',
  },
  {
    label: 'months',
    value: 'MONTHS',
  },
];
