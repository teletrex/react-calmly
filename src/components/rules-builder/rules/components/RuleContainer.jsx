/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import React, { useMemo, useState } from 'react';
import PropTypes from 'prop-types';
import isEmpty from 'lodash/isEmpty';

import { RULE_INPUT_TYPES, RULE_TYPES } from '../../constants';
import { getRuleContainerConfigData, getRuleId, getRuleOutput } from '../helpers';
import {
  ALL_AVAILABLE_OPERATORS,
  CUSTOM_RULE_DEFAULT_OPERATORS,
  DATE_RULE_OPERATORS,
  DEFAULT_PROPS,
  NUMBER_RULE_OPERATORS,
  PROPTYPES,
  SELECT_RULE_OPERATORS,
  STRING_RULE_OPERATORS,
  VALUES,
} from '../constants';

import RuleOperatorDropdown from './OperatorDropdown';
import TimeframeSelector from './TimeframeSelector';

const RULE_OPERATORS = {
  [RULE_TYPES.date]: DATE_RULE_OPERATORS,
  [RULE_TYPES.number]: NUMBER_RULE_OPERATORS,
  [RULE_TYPES.select]: SELECT_RULE_OPERATORS,
  [RULE_TYPES.string]: STRING_RULE_OPERATORS,
  [RULE_TYPES.custom]: {
    DEFAULT: CUSTOM_RULE_DEFAULT_OPERATORS,
    ALL: ALL_AVAILABLE_OPERATORS,
  },
};

const getOperatorsList = ruleConfig => {
  if (ruleConfig.type === RULE_TYPES.custom) {
    const customRuleOperators = isEmpty(ruleConfig.operators) ? 'DEFAULT' : 'ALL';
    return RULE_OPERATORS[RULE_TYPES.custom][customRuleOperators];
  }

  return RULE_OPERATORS[ruleConfig.type] || [];
};

const RuleContainer = ({ config, data, onRuleChange, timeframe, ruleComponent: RuleComponent }) => {
  const { operator, operators, value } = getRuleContainerConfigData(
    config,
    data,
    getOperatorsList(config)
  );

  const [isTimeframeVisible, setIsTimeframeVisible] = useState(!isEmpty(timeframe));
  const id = useMemo(() => getRuleId(config.type), [config.type]);

  const handleOperatorChange = ({ selectedItem: newOperator }) => {
    const newValue =
      newOperator.inputType === RULE_INPUT_TYPES.NONE ? VALUES.BLANK : VALUES.UNPROVIDED;

    onRuleChange(getRuleOutput(newOperator.value, newValue));
  };

  const handleValueChange = newValue => {
    onRuleChange(getRuleOutput(operator.value, newValue));
  };

  const setTimeframeVisibility = isVisible => {
    // allow to control the visibility only if hasTimeframe param was set
    if (!isEmpty(timeframe)) {
      if (!isVisible && !isEmpty(timeframe?.data)) {
        // reset timeframe if it has a value and is being hidden
        timeframe.onChange({});
      }
      setIsTimeframeVisible(isVisible);
    }
  };

  const commonProps = {
    id,
    operator,
  };

  return (
    <>
      <RuleOperatorDropdown
        onOperatorChange={handleOperatorChange}
        operators={operators}
        {...commonProps}
      />
      <RuleComponent
        config={config}
        onValueChange={handleValueChange}
        setTimeframeVisibility={setTimeframeVisibility}
        timeframe={timeframe}
        value={value}
        {...commonProps}
      />
      {isTimeframeVisible && <TimeframeSelector {...commonProps} timeframe={timeframe} />}
    </>
  );
};

RuleContainer.propTypes = {
  config: PropTypes.shape({
    operators: PROPTYPES.OPERATORS.isRequired,
    type: PropTypes.oneOf(Object.values(RULE_TYPES)).isRequired,
  }).isRequired,
  data: PropTypes.shape({
    operator: PROPTYPES.OPERATOR_NAME,
  }),
  onRuleChange: PropTypes.func.isRequired,
  ruleComponent: PropTypes.func.isRequired,
  timeframe: PROPTYPES.TIMEFRAME,
};

RuleContainer.defaultProps = {
  data: {
    operator: null,
  },
  timeframe: DEFAULT_PROPS.TIMEFRAME,
};

export default RuleContainer;
