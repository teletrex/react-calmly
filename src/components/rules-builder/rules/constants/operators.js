/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import { RULE_INPUT_TYPES } from '../../constants';

const {
  MULTIPLE_VALUES,
  NONE,
  PERIOD_NUMBER,
  PERIOD_RANGE,
  RANGE,
  SINGLE_VALUE,
  WHEN_RELATION,
  WITHIN_PERIOD,
} = RULE_INPUT_TYPES;

// Common operators
export const BLANK_OPERATORS = [
  {
    inputType: NONE,
    label: 'is blank',
    value: 'IS_BLANK',
  },
  {
    inputType: NONE,
    label: 'is not blank',
    value: 'IS_NOT_BLANK',
  },
];

export const EXACT_OPERATORS = [
  {
    inputType: WHEN_RELATION,
    label: 'is exactly',
    value: 'IS_EXACTLY',
  },
  {
    inputType: WHEN_RELATION,
    label: 'is not exactly',
    value: 'IS_NOT_EXACTLY',
  },
];

export const IS_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'is',
    value: 'IS',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'is not',
    value: 'IS_NOT',
  },
];

export const HAS_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'has',
    value: 'HAS',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'has not',
    value: 'HAS_NOT',
  },
];

export const IS_ONE_OF_OPERATORS = [
  {
    inputType: MULTIPLE_VALUES,
    label: 'is one of the following',
    value: 'IS_ONE_OF_THE_FOLLOWING',
  },
  {
    inputType: MULTIPLE_VALUES,
    label: 'is not one of the following',
    value: 'IS_NOT_ONE_OF_THE_FOLLOWING',
  },
];

export const CONTAIN_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'contains',
    value: 'CONTAINS',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'does not contain',
    value: 'DOES_NOT_CONTAIN',
  },
];

export const CONTAIN_ONE_OF_OPERATORS = [
  {
    inputType: MULTIPLE_VALUES,
    label: 'contains one of the following',
    value: 'CONTAINS_ONE_OF_THE_FOLLOWING',
  },
  {
    inputType: MULTIPLE_VALUES,
    label: 'does not contain one of the following',
    value: 'DOES_NOT_CONTAIN_ONE_OF_THE_FOLLOWING',
  },
];

export const RANGE_OPERATORS = [
  {
    inputType: RANGE,
    label: 'is between',
    value: 'IS_BETWEEN',
  },
];

// String related operators
export const STARTS_WITH_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'starts with',
    value: 'STARTS_WITH',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'does not start with',
    value: 'DOES_NOT_START_WITH',
  },
];

export const STARTS_WITH_ONE_OF_OPERATORS = [
  {
    inputType: MULTIPLE_VALUES,
    label: 'starts with one of the following',
    value: 'STARTS_WITH_ONE_OF_THE_FOLLOWING',
  },
  {
    inputType: MULTIPLE_VALUES,
    label: 'does not start with one of the following',
    value: 'DOES_NOT_START_WITH_ONE_OF_THE_FOLLOWING',
  },
];

export const ENDS_WITH_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'ends with',
    value: 'ENDS_WITH',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'does not end with',
    value: 'DOES_NOT_END_WITH',
  },
];

export const ENDS_WITH_ONE_OF_OPERATORS = [
  {
    inputType: MULTIPLE_VALUES,
    label: 'ends with one of the following',
    value: 'ENDS_WITH_ONE_OF_THE_FOLLOWING',
  },
  {
    inputType: MULTIPLE_VALUES,
    label: 'does not end with one of the following',
    value: 'DOES_NOT_END_WITH_ONE_OF_THE_FOLLOWING',
  },
];

// Number related operators
export const LESS_THAN_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'is less than',
    value: 'IS_LESS',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'is less than or equal to',
    value: 'IS_LESS_OR_EQUAL',
  },
];

export const GREATER_THAN_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'is greater than',
    value: 'IS_GREATER',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'is greater than or equal to',
    value: 'IS_GREATER_OR_EQUAL',
  },
];

// Date related operators
export const TODAY_OPERATORS = [
  {
    inputType: NONE,
    label: 'is today',
    value: 'IS_TODAY',
  },
  {
    inputType: NONE,
    label: 'is after today',
    value: 'IS_AFTER_TODAY',
  },
  {
    inputType: NONE,
    label: 'is not after today',
    value: 'IS_NOT_AFTER_TODAY',
  },
  {
    inputType: NONE,
    label: 'is before today',
    value: 'IS_BEFORE_TODAY',
  },
  {
    inputType: NONE,
    label: 'is not before today',
    value: 'IS_NOT_BEFORE_TODAY',
  },
  {
    inputType: NONE,
    label: 'anniversary is today',
    value: 'ANNIVERSARY_IS_TODAY',
  },
];

export const ANNIVERSARY_OPERATORS = [
  {
    inputType: WHEN_RELATION,
    label: 'anniversary is',
    value: 'ANNIVERSARY_IS',
  },
];

export const WITHIN_OPERATORS = [
  {
    inputType: WITHIN_PERIOD,
    label: 'is within the last',
    value: 'IS_WITHIN_THE_LAST',
  },
  {
    inputType: WITHIN_PERIOD,
    label: 'is not within the last',
    value: 'IS_NOT_WITHIN_THE_LAST',
  },
  {
    inputType: WITHIN_PERIOD,
    label: 'is within the next',
    value: 'IS_WITHIN_THE_NEXT',
  },
  {
    inputType: WITHIN_PERIOD,
    label: 'is not within the next',
    value: 'IS_NOT_WITHIN_THE_NEXT',
  },
];

export const WITHIN_RANGE_OPERATORS = [
  {
    inputType: PERIOD_RANGE,
    label: 'is within the range of the last',
    value: 'IS_WITHIN_THE_RANGE_OF_THE_LAST',
  },
];

export const PERIOD_EQUAL_OPERATORS = [
  {
    inputType: PERIOD_NUMBER,
    label: 'year is',
    value: 'YEAR_IS',
  },
  {
    inputType: PERIOD_NUMBER,
    label: 'month is',
    value: 'MONTH_IS',
  },
  {
    inputType: PERIOD_NUMBER,
    label: 'day is',
    value: 'DAY_IS',
  },
];

export const BEFORE_AFTER_DATE_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'is after',
    value: 'IS_AFTER',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'is before',
    value: 'IS_BEFORE',
  }
]

export const DATA_AGGREGATION_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'by',
    value: 'BY',
  },

];
