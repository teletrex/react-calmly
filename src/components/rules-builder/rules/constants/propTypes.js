/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import PropTypes from 'prop-types';

import { RULE_INPUT_TYPES } from '../../constants';

const { any, arrayOf, func, oneOf, oneOfType, shape, string } = PropTypes;

const OPERATOR = oneOfType([
  string,
  shape({
    inputType: oneOf(Object.values(RULE_INPUT_TYPES)).isRequired,
    label: string.isRequired,
    value: string.isRequired,
  }),
]);
const OPERATORS = arrayOf(OPERATOR);

const TIMEFRAME = shape({
  data: shape({
    operator: string,
    value: any,
  }),
  onChange: func,
});

const OPERATOR_NAME = string;

const SELECT_OPTION = shape({
  label: string.isRequired,
  value: string.isRequired,
});

const SELECT_OPTIONS = arrayOf(SELECT_OPTION);

export const PROPTYPES = {
  OPERATOR,
  OPERATORS,
  OPERATOR_NAME,
  SELECT_OPTION,
  SELECT_OPTIONS,
  TIMEFRAME,
};
