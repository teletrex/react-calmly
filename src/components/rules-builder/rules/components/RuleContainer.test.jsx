/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import React from 'react';
import { fireEvent, render, waitFor, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { RULE_TYPES } from '../../constants';
import { IS_OPERATORS, RANGE_OPERATORS, BLANK_OPERATORS } from '../constants/operators';
import { NUMBER_RULE_OPERATORS, STRING_RULE_OPERATORS, VALUES } from '../constants';

import RuleContainer from './RuleContainer';

import { CustomRule, StringRule, NumberRule } from '..';

describe('RulesContainer', () => {
  const defaultProps = {
    config: {
      operators: STRING_RULE_OPERATORS,
      type: RULE_TYPES.string,
      placeholder: {
        textInput: 'Text placeholder',
      },
    },
    onRuleChange: jest.fn(),
    ruleComponent: StringRule,
  };

  beforeEach(() => {
    beforeEach(() => jest.resetAllMocks());
  });

  it('should render correctly', () => {
    const { container } = render(<RuleContainer {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  it('should properly render a CustomRule with custom operators', () => {
    const props = {
      ...defaultProps,
      config: {
        operators: [{ label: 'test operator', value: 'TEST_OPERATOR', inputType: 'singleValue' }],
        type: RULE_TYPES.custom,
      },
      ruleComponent: CustomRule,
    };
    const { getByText, getByRole } = render(<RuleContainer {...props} />);

    const dropdown = getByRole('listbox');
    const dropdownButton = within(dropdown).getByRole('button');

    userEvent.click(dropdownButton);

    expect(getByText('test operator')).toBeInTheDocument();
  });

  it('should properly render a CustomRule with default operators only', () => {
    const props = {
      ...defaultProps,
      config: {
        operators: [],
        type: RULE_TYPES.custom,
      },
      ruleComponent: CustomRule,
    };
    const { getByText, queryByText, getByRole } = render(<RuleContainer {...props} />);

    const dropdown = getByRole('listbox');
    const dropdownButton = within(dropdown).getByRole('button');
    userEvent.click(dropdownButton);

    expect(getByText('has not')).toBeInTheDocument();
    expect(getByText('has')).toBeInTheDocument();
    expect(queryByText('is blank')).not.toBeInTheDocument();
  });

  it('should properly call onRuleChange after changing the operator', () => {
    const { getByRole, getByText } = render(<RuleContainer {...defaultProps} />);

    // expand the operator dropdown:
    const dropdown = getByRole('listbox');
    const dropdownButton = within(dropdown).getByRole('button');
    userEvent.click(dropdownButton);

    // select a new operator:
    userEvent.click(getByText(STRING_RULE_OPERATORS[1].label));

    expect(defaultProps.onRuleChange).toHaveBeenCalledWith({
      operator: STRING_RULE_OPERATORS[1].value,
    });
  });

  it('should properly handle blank values', () => {
    const { getByRole, getByText } = render(<RuleContainer {...defaultProps} />);

    // expand the operator dropdown:
    const dropdown = getByRole('listbox');
    const dropdownButton = within(dropdown).getByRole('button');
    userEvent.click(dropdownButton);

    // select a new operator:
    userEvent.click(getByText(BLANK_OPERATORS[0].label));

    expect(defaultProps.onRuleChange).toHaveBeenCalledWith({
      operator: BLANK_OPERATORS[0].value,
      value: VALUES.BLANK,
    });
  });

  it('should properly call onRuleChange after changing the value', () => {
    const props = {
      ...defaultProps,
      data: {
        operator: STRING_RULE_OPERATORS[0].value,
        value: 'Initial value',
      },
    };
    const { getByPlaceholderText } = render(<RuleContainer {...props} />);
    const textInput = getByPlaceholderText(defaultProps.config.placeholder.textInput);

    fireEvent.change(textInput, { target: { value: 'New value' } });
    fireEvent.blur(textInput);

    expect(defaultProps.onRuleChange).toHaveBeenCalledWith({
      operator: props.data.operator,
      value: 'New value',
    });
  });

  it('should call onRuleChange with formatted after changing the number value in NumberRule', async () => {
    const props = {
      ...defaultProps,
      config: {
        operators: NUMBER_RULE_OPERATORS,
        type: RULE_TYPES.number,
      },
      data: {
        operator: IS_OPERATORS[0].value,
        value: 3,
      },
      ruleComponent: NumberRule,
    };
    const { getByPlaceholderText } = render(<RuleContainer {...props} />);
    const numberInput = getByPlaceholderText('Enter value');

    fireEvent.change(numberInput, { target: { value: '2' } });
    fireEvent.blur(numberInput);
    await waitFor(() =>
      expect(defaultProps.onRuleChange).toHaveBeenCalledWith({
        operator: props.data.operator,
        value: 2,
      })
    );
  });

  it('should call onRuleChange with formatted value after changing the range value in NumberRule', async () => {
    const props = {
      ...defaultProps,
      config: {
        operators: NUMBER_RULE_OPERATORS,
        type: RULE_TYPES.number,
      },
      data: {
        operator: RANGE_OPERATORS[0].value,
        value: [3, 7],
      },
      ruleComponent: NumberRule,
    };
    const { getAllByPlaceholderText } = render(<RuleContainer {...props} />);
    const numberInputs = getAllByPlaceholderText('Enter value');

    fireEvent.change(numberInputs[0], { target: { value: '2' } });
    fireEvent.blur(numberInputs[0]);
    await waitFor(() =>
      expect(defaultProps.onRuleChange).toHaveBeenCalledWith({
        operator: props.data.operator,
        value: [2, 7],
      })
    );

    fireEvent.change(numberInputs[1], { target: { value: '9' } });
    fireEvent.blur(numberInputs[1]);
    await waitFor(() =>
      expect(defaultProps.onRuleChange).toHaveBeenCalledWith({
        operator: props.data.operator,
        value: [2, 9],
      })
    );
  });

  it('should render TimeframeSelector if provided with timeframe data', () => {
    const props = {
      ...defaultProps,
      timeframe: {
        operator: 'WITHIN_THE_LAST',
        value: {
          value: 7,
          period: 'DAYS',
        },
      },
    };
    const { getByTestId } = render(<RuleContainer {...props} />);

    expect(getByTestId('timeframe-selector')).toBeInTheDocument();
  });

  it('should properly set TimeframeSelector visibility if timeframe is enabled', () => {
    const InputComponent = ({ setTimeframeVisibility }) => (
      <button onClick={() => setTimeframeVisibility(false)} type="button">
        Hide timeframe
      </button>
    );
    const props = {
      ...defaultProps,
      config: {
        operators: [],
        type: RULE_TYPES.custom,
        inputComponent: InputComponent,
      },
      ruleComponent: CustomRule,
      timeframe: {
        onChange: jest.fn(),
        data: {
          operator: 'WITHIN_THE_LAST',
          value: {
            value: 7,
            period: 'DAYS',
          },
        },
      },
    };
    const { getByTestId, getByText, queryByTestId } = render(<RuleContainer {...props} />);
    expect(getByTestId('timeframe-selector')).toBeInTheDocument();

    fireEvent.click(getByText('Hide timeframe'));
    expect(queryByTestId('timeframe-selector')).not.toBeInTheDocument();
  });

  it('should properly reset the timeframe value on visibility change', () => {
    const InputComponent = ({ setTimeframeVisibility }) => (
      <button onClick={() => setTimeframeVisibility(false)} type="button">
        Hide timeframe
      </button>
    );
    const props = {
      ...defaultProps,
      config: {
        operators: [],
        type: RULE_TYPES.custom,
        inputComponent: InputComponent,
        hasTimeframe: true,
      },
      ruleComponent: CustomRule,
      timeframe: {
        onChange: jest.fn(),
        data: {
          operator: 'WITHIN_THE_LAST',
          value: {
            value: 7,
            period: 'DAYS',
          },
        },
      },
    };
    const { getByDisplayValue, getByText } = render(<RuleContainer {...props} />);
    expect(getByDisplayValue('7')).toBeInTheDocument();

    fireEvent.click(getByText('Hide timeframe'));
    expect(props.timeframe.onChange).toHaveBeenCalledWith({});
  });

  it("should not reset the timeframe value if it's empty", () => {
    const InputComponent = ({ setTimeframeVisibility }) => (
      <button onClick={() => setTimeframeVisibility(false)} type="button">
        Hide timeframe
      </button>
    );
    const props = {
      ...defaultProps,
      config: {
        operators: [],
        type: RULE_TYPES.custom,
        inputComponent: InputComponent,
        hasTimeframe: true,
      },
      ruleComponent: CustomRule,
      timeframe: {
        onChange: jest.fn(),
        data: {},
      },
    };
    const { getByText } = render(<RuleContainer {...props} />);

    fireEvent.click(getByText('Hide timeframe'));
    expect(props.timeframe.onChange).not.toHaveBeenCalledWith();
  });

  it('should not allow to show the TimeframeSelector if timeframe is not enabled', () => {
    const InputComponent = ({ setTimeframeVisibility }) => (
      <button onClick={() => setTimeframeVisibility(true)} type="button">
        Show timeframe
      </button>
    );
    const props = {
      ...defaultProps,
      config: {
        operators: [],
        type: RULE_TYPES.custom,
        inputComponent: InputComponent,
      },
      ruleComponent: CustomRule,
    };
    const { getByText, queryByTestId } = render(<RuleContainer {...props} />);
    expect(queryByTestId('timeframe-selector')).not.toBeInTheDocument();

    fireEvent.click(getByText('Show timeframe'));
    expect(queryByTestId('timeframe-selector')).not.toBeInTheDocument();
  });

  it('should set the first operator if disableCriteriaReselect is switched on', () => {
    const props = {
      ...defaultProps,
      config: {
        operators: [],
        type: RULE_TYPES.string,
        disableCriteriaReselect: true,
      },
      ruleComponent: StringRule,
    };

    const { getAllByText, getByRole } = render(<RuleContainer {...props} />);

    const dropdown = getByRole('listbox');
    const dropdownButton = within(dropdown).getByRole('button');
    userEvent.click(dropdownButton);
    expect(getAllByText(STRING_RULE_OPERATORS[0].label)).toHaveLength(2);
  });
});
