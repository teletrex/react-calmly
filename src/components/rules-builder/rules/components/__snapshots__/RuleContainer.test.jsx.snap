// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RulesContainer should render correctly 1`] = `
<div>
  <div
    class="bx-proposal-dropdown-wrapper bx-proposal-dropdown-wrapper--inline"
  >
    <div
      class="bx-proposal-dropdown"
      role="listbox"
    >
      <span>
        <div
          aria-label="Select operator"
          class="bx-proposal-dropdown__field"
          role="button"
          tabindex="0"
        >
          <span
            class="bx-proposal-dropdown__field__label"
          >
            Select operator
          </span>
          <div
            class="bx-proposal-dropdown__field__menu-icon"
          >
            <svg
              aria-hidden="true"
              focusable="false"
              height="16"
              preserveAspectRatio="xMidYMid meet"
              style="will-change: transform;"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 11L3 6l.7-.7L8 9.6l4.3-4.3.7.7z"
              />
            </svg>
          </div>
        </div>
      </span>
    </div>
  </div>
</div>
`;
